torchao/dtypes/nf4tensor.py

```diff
@@ -698,7 +699,7 @@ def quantize_tensor_nearest(
         value = value.unsqueeze(-1)  # (numel, 1)
         # Compare the value tensor with the nf4 tensor element-wise
         diff = (value - nf4).abs()
-        closest_nf4 = diff.min(dim=-1).indices
+        closest_nf4 = diff.min(dim=-1).indices.to(dtype=torch.uint8)
```

Comment:
On another note, it seems like we could use 4 bits here? We have https://github.com/pytorch/ao/blob/03e2c9b056a17a0e6d5ae31c2a73df36345f886f/torchao/dtypes/uint4.py , but it likely needs some work. I assume `closest_nf4` here is an index into the nf4 table which by definition only is 4 bits and thus there aren't more than 4bit worth of indices?

Reply:
Bit-packing logic is done in `convert_to_norm_float_weight()`. I don't know the original author's intention but one benefit of not doing bit-packing in `quantize_tensor_nearest()` is that we can do correctness test (without unpacking bits).

Reply:
Mhm, I see. So then using compile at a higher level might still avoid materializing the full int8 tensor (thus reducing peak memory utilization during construction).

Reply:
Also, from what I remember when I wrote this, the bitshifting is only supported on uint8 dtype: https://github.com/pytorch/ao/blob/03e2c9b056a17a0e6d5ae31c2a73df36345f886f/torchao/dtypes/nf4tensor.py#L661